products/libllbuild/BuildDB-C-API.cpp

```diff
@@ -0,0 +1,286 @@
+//
+//  BuildDB-C-API.c
+//  libllbuild
+//
+//  Copyright © 2019 Apple Inc. All rights reserved.
+//
+
+#include <llbuild/llbuild.h>
+
+#include "llbuild/Core/BuildDB.h"
+
+#include "llvm/Support/raw_ostream.h"
+
+#include <algorithm>
+#include <cassert>
+#include <memory>
+
+using namespace llbuild;
+using namespace llbuild::core;
+
+namespace {
+
+class CAPIBuildDBResultKeys {
+private:
+  std::unique_ptr<std::map<KeyID, KeyType>> keys;
+  std::unique_ptr<std::map<KeyType, KeyID>> ids;
+  
+public:
+  CAPIBuildDBResultKeys(std::map<KeyID, KeyType> keys): keys(llvm::make_unique<std::map<KeyID, KeyType>>(keys)) {
+    ids = llvm::make_unique<std::map<KeyType, KeyID>>();
+    for (const auto &pair : keys) {
+      (*ids.get())[std::get<1>(pair)] = std::get<0>(pair);
+    }
+  }
+  
+  CAPIBuildDBResultKeys(const CAPIBuildDBResultKeys&) LLBUILD_DELETED_FUNCTION;
+  CAPIBuildDBResultKeys& operator=(const CAPIBuildDBResultKeys&) LLBUILD_DELETED_FUNCTION;
+  
+  const uint32_t size() {
+    return keys.get()->size();
+  }
+  
+  const KeyType keyForID(KeyID index) {
+    return (*keys.get())[index];
+  }
+  
+  const KeyID idForKey(KeyType key) {
+    return (*ids.get())[key];
+  }
+};
+  
+class CAPIBuildDBDelegate: public BuildDBDelegate {
+  llb_database_delegate_t cAPIDelegate;
+  
+public:
+  CAPIBuildDBDelegate(llb_database_delegate_t delegate): cAPIDelegate(delegate) { }
+  
+  KeyType getKeyForID(KeyID keyID) override {
+    if (!cAPIDelegate.get_key_for_id) {
+      return nullptr;
+    }
+    
+    llb_database_key_type key;
+    cAPIDelegate.get_key_for_id(cAPIDelegate.context, keyID, &key);
+    
+    return std::string(key);
+  }
+  
+  KeyID getKeyID(const KeyType &key) override {
+    if (!cAPIDelegate.get_key_id) {
+      return 0;
+    }
+    
+    return cAPIDelegate.get_key_id(cAPIDelegate.context, (char *)key.c_str());
+  }
+};
+  
+class CAPIBuildDB {
+  CAPIBuildDBDelegate cAPIDelegate;
+  
+  std::unique_ptr<BuildDB> _db;
+  
+public:
+  CAPIBuildDB(StringRef path, uint32_t clientSchemaVersion, CAPIBuildDBDelegate delegate, std::string *error_out): cAPIDelegate(delegate) {
+    _db = createSQLiteBuildDB(path, clientSchemaVersion, error_out);
+    _db.get()->attachDelegate(&cAPIDelegate);
```

Comment:
This will crash if there is an error creating the database... that will need to be handled in some way

Reply:
Also, this typically would be written as just `_db->`

Reply:
Isn't the `get()` necessary because of the unique_ptr?